add mira220 image sensor

fiepfiep opened the pull request:
add driver / device tree support for ams [mira220](https://ams-osram.com/products/sensor-solutions/cmos-image-sensors/ams-mira220)

needed changes also done for [libcamera](https://github.com/raspberrypi/libcamera/pull/248).

6by9 commented on arch/arm/boot/dts/overlays/mira220-overlay.dts:
```diff
@@ -0,0 +1,102 @@
+		};
+	};
+
+	fragment@202 {
```
Fragment is dormant and not referenced by an override.

fiepfiep replied:
done, has been removed.

6by9 commented on arch/arm/boot/dts/overlays/mira220-overlay.dts:
```diff
@@ -0,0 +1,102 @@
+		};
+	};
+
+	fragment@201 {
```
Fragment is dormant and not referenced by an override.

fiepfiep replied:
resolved.

6by9 commented on arch/arm/boot/dts/overlays/mira220.dtsi:
```diff
@@ -0,0 +1,40 @@
+			data-lanes = <1 2>;
+			// clock-noncontinuous;
+			link-frequencies =
+				/bits/ 64 <456000000>;
```
Is this the correct link frequency for this sensor?

fiepfiep replied:
currently operating at 1.5 Gbit/s per lane

fiepfiep replied:
solved, set to 750, as datarate is 1500Mbit/s

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,4366 @@
+ 
+ /* Embedded metadata stream structure */
+ #define MIRA220_EMBEDDED_LINE_WIDTH 16384
+ #define MIRA220_NUM_EMBEDDED_LINES 1
```
Does this sensor actually have useful embedded metadata? It's disabled in libcamera.

I'd made the same comment in https://github.com/raspberrypi/linux/pull/6552#discussion_r1893800072 that this isn't how mainline are implementing metadata, so it's simpler to not implement it in the driver if not needed.

fiepfiep replied:
I have removed the embedded data

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,4366 @@
+ // 1600_1400_30fps_12b_2lanes
+
+ static const struct mira220_reg full_1600_1400_1500_12b_2lanes_reg[] = {
+	{0x1003,0x2}, //,Initial Upload
```
Comment isn't useful.

fiepfiep replied:
solved

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,4366 @@
+ 
+ // 1600_1400_30fps_12b_2lanes
+
+ static const struct mira220_reg full_1600_1400_1500_12b_2lanes_reg[] = {
```
You have about 18 registers that differ between these 3 huge tables. Break out all the common stuff into a common table, and only send the bits that change per mode

fiepfiep replied:
thanks, will do.

fiepfiep replied:
simplified, reduced driver to 1 mode only.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,4366 @@
+ #define MIRA220_PIXEL_ARRAY_WIDTH		1600U
+ #define MIRA220_PIXEL_ARRAY_HEIGHT		1400U
+ 
+ #define MIRA220_ANALOG_GAIN_REG			0x400A
```
Comment made at https://github.com/raspberrypi/linux/pull/6552#discussion_r1893820023 that the V4L2_CCI is preferred rather than bare register defines as it encodes the size of each register.

fiepfiep replied:
I converted everything to the new style cci writes

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,4366 @@
+ static const char * const mira220_supply_name[] = {
+	 // TODO(jalv): Check supply names
+	 /* Supplies can be enabled in any order */
+	 "VANA",  /* Analog (2.8V) supply */
```
Regulators should be named in lower case.

fiepfiep replied:
done.

6by9 reviewed (changes requested):
I've started to review this, but I'm noticing a lot of my comments from #6552 haven't been addressed.

If embedded data isn't required, then do not include it.

pelwell commented:
I'll warn you now that, once the content of these commits has been approved, the end result should ideally be 3 commits - driver changes, config changes, and overlay/README changes.

pelwell commented on arch/arm/boot/dts/overlays/mira220-overlay.dts:
```diff
@@ -58,28 +58,6 @@
 		};
 	};
 
-	fragment@102 {
```
Did you delete this fragment by mistake? The `media-controller` parameter refers to it.

fiepfiep replied:
Yes that is a mistake.

fiepfiep replied:
I put it back. thanks for finding out.

fiepfiep commented:
> I'll warn you now that, once the content of these commits has been approved, the end result should ideally be 3 commits - driver changes, config changes, and overlay/README changes.

I will keep it in mind. Once all is approved by you I will refactor it to 3 commits.

fiepfiep commented:
I cleaned up the driver a bit, implemented the new CCI_REG style writes, and took care of 6by9's comments.
Please let me know if you find other issues before we can do the merge.

pelwell commented on arch/arm/boot/dts/overlays/README:
```diff
@@ -3632,6 +3649,7 @@ Params: rotation                Mounting rotation of the camera sensor (0 or
                                 Appears not to work on Pi3.
 
 
+
```
This looks like an unnecessary change.

pelwell commented on arch/arm/boot/dts/overlays/README:
```diff
@@ -3511,6 +3511,23 @@ Params:
                                 (default 16).
 
 
+
```
2 blank lines is enough.

pelwell commented on arch/arm/boot/dts/overlays/README:
```diff
@@ -3511,6 +3511,23 @@ Params:
+                                configuring the sensor (default on)
+        cam0                    Adopt the default configuration for CAM0 on a
+                                Compute Module (CSI0, i2c_vc, and cam0_reg).
+
```
And here.

pelwell reviewed (commented):
Apart from the extra blank lines in the README, the overlay looks good.

Checkpatch has lots of complaints about the driver which you should really fix. Nobody using `printk` directly anymore, for example.

The only additional change I would like (apart from refactoring down to 3 commits as discussed) is the addition of CONFIG_VIDEO_MIRA220 to the new PREEMPT_RT config file arm64/bcm2711_rt_defconfig.

6by9 commented on arch/arm/boot/dts/overlays/mira220-overlay.dts:
```diff
@@ -0,0 +1,87 @@
+		       <&csi_frag>, "target:0=",<&csi0>,
+		       <&clk_frag>, "target:0=",<&cam0_clk>,
+		       <&cam_node>, "clocks:0=",<&cam0_clk>,
+		       <&cam_node>, "VANA-supply:0=",<&cam0_reg>;
```
Your supply is now `vana-supply`, so this won't match.

6by9 commented on drivers/media/i2c/Kconfig:
```diff
@@ -149,6 +149,20 @@ config VIDEO_HI847
           To compile this driver as a module, choose M here: the
           module will be called hi847.
 
+config VIDEO_MIRA220
```
These are normally in alphabetical order. (I see the Arducam drivers have got out of position during forward porting).

Annoyingly this will also move the position in the defconfigs, so perhaps I don't care enough.

6by9 commented on drivers/media/i2c/Makefile:
```diff
@@ -49,6 +49,7 @@ obj-$(CONFIG_VIDEO_HI556) += hi556.o
 obj-$(CONFIG_VIDEO_HI846) += hi846.o
 obj-$(CONFIG_VIDEO_HI847) += hi847.o
 obj-$(CONFIG_VIDEO_I2C) += video-i2c.o
+obj-$(CONFIG_VIDEO_MIRA220)	+= mira220.o
```
Ditto - alphabetical order please.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2481 @@
+
+#define MIRA220_ANALOG_GAIN_MIN 1
+#define MIRA220_ANALOG_GAIN_MAX \
+	1 /* Fixed analog gain to 1, to avoid unexpected behavior. */
```
Weird line breaking. That does seem quite a limitation though if you have no analogue gain.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2481 @@
+// Outdated. See below.
+// pixel_rate = link_freq * 2 * nr_of_lanes / bits_per_sample
+// 1.0Gb/s * 2 * 2 / 12 = 357913941
+// #define MIRA220_PIXEL_RATE		(357913941)
```
If it's outdated, then please remove it.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2481 @@
+#define MIRA220_PIXEL_ARRAY_LEFT 0U
+#define MIRA220_PIXEL_ARRAY_TOP 0U
+#define MIRA220_PIXEL_ARRAY_WIDTH 1600U
+#define MIRA220_PIXEL_ARRAY_HEIGHT 1400U
```
Looking at https://look.ams-osram.com/m/2feacfd5e2a2d20e/original/Mira220-1-2-7-2-2MP-NIR-Enhanced-Global-Shutter-Image.pdf, Figure 1 says the sensor has 32 extra rows top and bottom, and 21 extra columns left and right.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2481 @@
+
+/* Embedded metadata stream structure */
+#define MIRA220_EMBEDDED_LINE_WIDTH 16384
+#define MIRA220_NUM_EMBEDDED_LINES 1
```
You'd said in https://github.com/raspberrypi/linux/pull/6717#discussion_r2025643185 "I have removed the embedded data", so why are these still here. Likewise METADATA_PAD at line 158.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2481 @@
+};
+
+static const struct cci_reg_sequence full_1600_1400_1500_12b_2lanes_reg_new[] = {
+	{ CCI_REG8(0x1003), 0x2 }, //,Initial Upload
```
This comment of `//,Initial Upload` is still present and unnecessary.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2481 @@
+  * HBLANK=1/((1/204813108)*30*(1400+13))-1600=3232
+  */
+
+#define MIRA220_HBLANK_640x480_120FPS 3860
```
You've reduced the driver to one mode, so most of these `MIRA220_HBLANK_` defines are redundant.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2481 @@
+static const char *const mira220_supply_name[] = {
+	// TODO(jalv): Check supply names
+	/* Supplies can be enabled in any order */
+	"VANA", /* Analog (2.8V) supply */
```
Still in capitals, and that doesn't match the dtoverlay.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2481 @@
+	for (try_cnt = 0; try_cnt < 5; try_cnt++) {
+		ret = cci_write(mira220->regmap, MIRA220_IMAGER_STATE_REG,
+				MIRA220_IMAGER_STATE_STOP_AT_ROW, NULL);
+		if (ret) {
```
You really have such dodgy I2C comms such that you have to retry?

fiepfiep replied:
probably no longer needed indeed. There used to be a problem but this was resolved.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2481 @@
+
+// Returns the maximum exposure time in row_length (reg value).
+// Calculation is baded on Mira220 datasheet Section 9.2.
+static u32 mira220_calculate_max_exposure_time(u32 vsize, u32 vblank,
```
vsize, more commonly known as height.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2481 @@
+			break;
+
+	if (i >= ARRAY_SIZE(codes)) {
+		dev_err_ratelimited(&client->dev,
```
Error message not needed, and gives a route for userspace to be able to spam the kernel log (even if rate limited).

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2481 @@
+
+	/*test cci write*/
+	ret = cci_read(mira220->regmap, MIRA220_EXP_TIME_REG, &readval, &ret);
+	otherval = readval & 0xFFFF;
```
Why read the value before writing it? You do nothing with the result.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2481 @@
+	if (mira220_check_hwcfg(dev))
+		return -EINVAL;
+
+	/* Parse device tree to check if dtoverlay has param skip-reg-upload=1 */
```
Huh?

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2481 @@
+}
+
+static const struct dev_pm_ops mira220_pm_ops = {
+	SET_SYSTEM_SLEEP_PM_OPS(mira220_suspend, mira220_resume)
```
You shouldn't be implementing SLEEP_OPS.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2481 @@
+		SET_RUNTIME_PM_OPS(mira220_power_off, mira220_power_on, NULL)
+};
+
+static const struct of_device_id mira220_dt_ids[] = { { .compatible =
```
Please use the normal formatting for dt_ids.
```
static const struct of_device_id mira220_dt_ids[] = {
	{ .compatible = "ams,mira220" },
	{ /* sentinel */ }
};
```

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2481 @@
+MODULE_DEVICE_TABLE(of, mira220_dt_ids);
+
+static const struct i2c_device_id mira220_ids[] = { { "mira220", 0 }, {} };
+MODULE_DEVICE_TABLE(i2c, mira220_ids);
```
You shouldn't need I2C ids. This driver needs config data, so you can't load via sysfs.

6by9 reviewed (commented):
Still a number of my previous review comments haven't been addressed.

- Please remove the references to the embedded data pad.
- Please remove all commented out code.
- Please remove all redundant comments.
- Please rationalise the logging messages. Very few should be needed at all.
- Regulator names should be consistently in lower case.

pelwell commented:
FYI the build checks explicitly set CONFIG_WERROR=y, overriding the defconfigs:
```
scripts/config --file ${{github.workspace}}/build/.config --set-val CONFIG_WERROR y
```

fiepfiep commented:
@pelwell could you run your checks again? I split my changes into 3 commits as requested.

pelwell commented:
It looks like somebody has started them. Please mark the PR as "Ready for review" when you think it is. However, there is no excuse for some of these (https://github.com/raspberrypi/linux/actions/runs/14470568709/job/40583365956?pr=6717#step:6:80) checkpatch errors such as trailing whitespace.

Commit:
defconfigs: Add ams Mira220 image sensor to Pi defconfigs

Adds defconfig for the Mira220 1600x1400 global
shutter image sensor

Signed-off-by: philippe baetens <philippe.baetens@ams-osram.com>

Commit:
dtoverlays: Add overlay for ams Mira220 image sensor

Adds an overlay for the Mira220 1600x1400 global
shutter image sensor

Signed-off-by: philippe baetens <philippe.baetens@ams-osram.com>

Commit:
media/i2c: Add driver for ams Mira220

Adds a driver for the NIR-enhanced Mira220 1600x1400 global
shutter image sensor.

Signed-off-by: philippe baetens <philippe.baetens@ams-osram.com>

fiepfiep commented:
I still get this error in checkpatch, what does it mean?
```
 WARNING: From:/Signed-off-by: email name mismatch: 'From: philippe baetens <philippe.baetens@ams-osram.com>' != 'Signed-off-by: Philippe Baetens <philippe.baetens@ams-osram.com>'
Error: WARNING: From:/Signed-off-by: email name mismatch: 'From: philippe baetens <philippe.baetens@ams-osram.com>' != 'Signed-off-by: Philippe Baetens <philippe.baetens@ams-osram.com>'
```

pelwell commented:
Is it possibly just that your name appears with and without uppercase letters? I'm not bothered about that, but it makes sense to fix it to avoid such errors in the future.

fiepfiep commented:
> Is it possibly just that your name appears with and without uppercase letters? I'm not bothered about that, but it makes sense to fix it to avoid such errors in the future.

that seems to be the case indeed. i modified the commit signoff message + some of the checkpatch remarks. Hope it is ok now.

pelwell commented:
That's fixed it. I'll give these builds time to complete, in which time 6by9 may want to chip in, but I've been taking his silence for consent.

6by9 commented on arch/arm/configs/bcm2709_defconfig:
```diff
@@ -934,6 +934,7 @@ CONFIG_VIDEO_BCM2835_UNICAM_LEGACY=m
 CONFIG_VIDEO_BCM2835_UNICAM=m
 CONFIG_VIDEO_ARDUCAM_64MP=m
 CONFIG_VIDEO_ARDUCAM_PIVARIETY=m
+CONFIG_VIDEO_MIRA220=m
```
With the move within Kconfig to keep everything in alphabetical order, I'm expecting this to be the wrong location in the defconfig.

fiepfiep replied:
right. will move to the line below the IMX sensors.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2160 @@
+	//MEDIA_BUS_FMT_Y8_1X8,
+	//MEDIA_BUS_FMT_Y10_1X10,
+	//MEDIA_BUS_FMT_Y12_1X12,
+	MEDIA_BUS_FMT_SGRBG8_1X8,
```
The comment just before this says that the sensor is monochrome, and yet you've used the GRBG Bayer pattern code. So which is it?

fiepfiep replied:
sensor comes both as RGB and mono. For now, I'm only doing the RGB.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2160 @@
+	reg_list = &mira220->mode->reg_list;
+	ret = cci_multi_reg_write(mira220->regmap, reg_list->regs,
+				  reg_list->num_of_regs, NULL);
+	// ret = mira220_write_regs(mira220, reg_list->regs, reg_list->num_of_regs);
```
Remove the now redundant line.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2160 @@
+	int ret = 0;
+
+	mutex_lock(&mira220->mutex);
+	if (mira220->streaming == enable) {
```
This should never be possible as the framework blocks it.

fiepfiep replied:
do you mean all mutex statements are not needed?

6by9 replied:
`mira220->streaming` is not required. The framework should never call `set_stream(..., 1)` if the stream is already enabled, or `set_stream(... , 0)` if there wasn't also a previous call to `set_stream(..., 1)` that returned 0.

If that means you don't need the mutex, then that's a bonus.

fiepfiep replied:
ah, clear. Will fix that too.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2160 @@
+	struct mira220 *mira220 = to_mira220(sd);
+	int ret;
+
+	if (mira220->streaming) {
```
resume shouldn't start streaming directly as the whole pipeline needs to be started correctly.

fiepfiep replied:
resume got deleted (unused)

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2160 @@
+
+	// Make the vblank control read only. This could be changed to allow changing framerate in
+	// runtime, but would require adapting other settings
+	// mira220->vblank->flags |= V4L2_CTRL_FLAG_READ_ONLY;
```
This has changed in that you have VBLANK control, so remove the comment and the commented out code.

fiepfiep replied:
sure.

fiepfiep commented:
So - what are the next steps to get this driver into the raspberry linux kernel?
At the moment I'm only merging into the 6.12 kernel. What about older or newer kernel versions? 
How does this driver propagate to official bookworm builds?

thanks.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2160 @@
+	(void)mira220;
+
+	regulator_bulk_disable(MIRA220_NUM_SUPPLIES, mira220->supplies);
+	clk_disable_unprepare(mira220->xclk);
```
Clock and regulators are disabled in the opposite order to being prepared.

fiepfiep replied:
ok, will swap order.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2160 @@
+	return ret;
+}
+
+static int __maybe_unused mira220_suspend(struct device *dev)
```
Both mira220_suspend and mira220_resume are totally unused. Remove them.

fiepfiep replied:
will remove them.

6by9 reviewed (changes requested):
In reading through to see if my previous comments had been addressed, add a number more.

There are a fair number of commented out blocks of code. They should be reviewed and removed where redundant.

6by9 commented:
> So - what are the next steps to get this driver into the raspberry linux kernel? At the moment I'm only merging into the 6.12 kernel. What about older or newer kernel versions? How does this driver propagate to official bookworm builds?

Older kernels won't get updated.

Later kernels will get a forward port, as long as there aren't any significant build changes required by the every-changing kernel APIs. Simple changes are likely to get fixed up. Significant changes without obvious fixups will be reported back to you to fix - we don't undertake the full support burden of 3rd party drivers, and they may get dropped from the defconfigs if not fixed and/or get the `BROKEN` tag added to the Kconfig entry if the situation persists for a significant period.

rpi-update gets kernel updates normally several time a week.
Raspberry Pi OS gets updates every few months based on when significant things occur, providing no significant issues are known about at that point.

pelwell commented:
Some of @6by9's comments have yet to be addressed, but the overlay and general structure look okay to me.

fiepfiep commented:
> Some of @6by9's comments have yet to be addressed, but the overlay and general structure look okay to me.

Thanks for reviewing. I believe I addressed all of @6by9's comments, but I might have missed something.. Can you be a bit more specific? thanks!

pelwell commented on arch/arm/boot/dts/overlays/mira220.dtsi:
```diff
@@ -0,0 +1,40 @@
+// Definitions for MIRA220 camera module on VC I2C bus
+
+
+//#include <dt-bindings/gpio/gpio.h>
```
Is this commented out section left here for a reason? If so, explain why in a comment, otherwise delete.

pelwell commented:
On this page (https://github.com/raspberrypi/linux/pull/6717/files), scroll down and expand mira.c, then look at the comments that remain, e.g.: https://github.com/raspberrypi/linux/pull/6717/files#r2027459784 and https://github.com/raspberrypi/linux/pull/6717/files#r2027467385

pelwell commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2087 @@
+#include <media/v4l2-cci.h>
+#include <linux/unaligned.h>
+
+/* Active pixel array is 1600 (H) x 1400 (V) pixels. 
```
Remove the space at the end of this line, squash the commits back to 3, then I think we're good to merge.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2087 @@
+	const struct cci_reg_sequence *regs;
+};
+
+struct mira220_v4l2_reg {
```
Unused.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2087 @@
+};
+
+static const struct cci_reg_sequence full_1600_1400_1500_12b_2lanes_reg_new[] = {
+	{ CCI_REG8(0x1003), 0x2 }, //  Initial Upload
```
All of the `// Initial upload` comments are meaningless

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2087 @@
+	struct v4l2_ctrl *gain;
+
+	u16 mira220_reg_w_cached_addr;
+	u8 mira220_reg_w_cached_flag;
```
Neither are used.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2087 @@
+		code->code = mira220_validate_format_code_or_default(
+			mira220, codes[code->index]);
+	} else {
+		if (code->index > 0)
```
This else clause is redundant with no metadata pad.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2087 @@
+	if (fse->pad >= NUM_PADS)
+		return -EINVAL;
+
+	if (fse->pad == IMAGE_PAD) {
```
You've already guaranteed this condition with the `if (fse->pad >= NUM_PADS)` check

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2087 @@
+		struct v4l2_mbus_framefmt *try_fmt =
+			v4l2_subdev_state_get_format(sd_state, fmt->pad);
+
+		try_fmt->code = fmt->pad == IMAGE_PAD ?
```
`fmt->pad` will be `IMAGE_PAD`, so no need for the ? : conditional

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2087 @@
+			__v4l2_ctrl_s_ctrl(mira220->vblank,
+					   mira220->mode->min_vblank);
+		}
+	} else {
```
There is no else.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2087 @@
+		goto err_rpm_put;
+	}
+
+	__v4l2_ctrl_grab(mira220->vflip, true);
```
Grabbing the flip controls is normally only needed if changing the flip changes the Bayer order. This driver appears not to do that.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2087 @@
+
+	mira220_otp_power_off(mira220);
+
+	return 0;
```
If we never fail the identity check, then make it a void function.

You never check any of the returned values from `mira220_otp_read` which would be a basic identity check.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2087 @@
+	mira220->hflip = v4l2_ctrl_new_std(ctrl_hdlr, &mira220_ctrl_ops,
+					   V4L2_CID_HFLIP, 0, 1, 1, 0);
+	if (mira220->hflip)
+		mira220->hflip->flags |= V4L2_CTRL_FLAG_MODIFY_LAYOUT;
```
In what way does it modify the layout? Normally this is when it changes the Bayer order, but your sensor appears not to do that.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2087 @@
+
+	ctrl_hdlr = &mira220->ctrl_handler;
+	/* v4l2_ctrl_handler_init gives a hint/guess of the number of v4l2_ctrl_new */
+	ret = v4l2_ctrl_handler_init(ctrl_hdlr, 16);
```
How many controls do you create? I make it 10, not 16.

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2087 @@
+		return ret;
+	}
+
+	usleep_range(1000000, 1000000 + 100);
```
Slightly crazy range of 100us tolerance on 1second.

`fsleep` is the normally recommended call now - https://www.kernel.org/doc/html/latest/timers/delay_sleep_functions.html

6by9 commented on drivers/media/i2c/mira220.c:
```diff
@@ -0,0 +1,2087 @@
+
+struct mira220 {
+	struct v4l2_subdev sd;
+	struct media_pad pad[NUM_PADS];
```
Hardly worth having an array with one element.

6by9 reviewed (commented):
There are still a number of remnants from the metadata pad kicking around. The things I've listed below would be nice to be cleaned up, but seeing as you'll be retaining responsibility for maintaining it, I'm not going to fuss. None of them should break the system.
You would be picked up on all these things if you try upstreaming the driver.

pelwell commented:
This still applies (both aspects):

> The only additional change I would like (apart from refactoring down to 3 commits as discussed) is the addition of CONFIG_VIDEO_MIRA220 to the new PREEMPT_RT config file arm64/bcm2711_rt_defconfig.

fiepfiep commented:
> This still applies (both aspects):
> 
> > The only additional change I would like (apart from refactoring down to 3 commits as discussed) is the addition of CONFIG_VIDEO_MIRA220 to the new PREEMPT_RT config file arm64/bcm2711_rt_defconfig.

okay, if all is good - I'll start refactoring down to 3 commits. I've added the rt_defconfig too.
